libs/ui/src/session_time_limit_tracker.tsx

```diff
@@ -0,0 +1,206 @@
+    // SECONDS_AHEAD_OF_OVERALL_SESSION_TIME_LIMIT_TO_WARN
+    secondsToSessionExpiry <=
+    SECONDS_AHEAD_OF_OVERALL_SESSION_TIME_LIMIT_TO_WARN
+  );
+}
+
+interface SessionTimeLimitTrackerHelperProps {
+  authStatus: AuthStatusWithSessionExpiry;
+  logOut: () => void;
+  updateSessionExpiry: (sessionExpiresAt: UnixTimestampInMilliseconds) => void;
+}
+
+/**
+ * A helper for SessionTimeLimitTracker
+ */
+function SessionTimeLimitTrackerHelper({
+  authStatus,
+  logOut,
+  updateSessionExpiry,
+}: SessionTimeLimitTrackerHelperProps): JSX.Element | null {
+  // TODO: Pull values from system settings
+  const inactiveSessionTimeLimitMinutes =
+    DEFAULT_INACTIVE_SESSION_TIME_LIMIT_MINUTES;
+  const overallSessionTimeLimitHours = DEFAULT_OVERALL_SESSION_TIME_LIMIT_HOURS;
+
+  const now = useNow().toJSDate();
+  const [
+    hasInactiveSessionTimeLimitBeenHit,
+    setHasInactiveSessionTimeLimitBeenHit,
+  ] = useState(false);
+  const [isModalDismissed, setIsModalDismissed] = useState(false);
+
+  function dismissModal() {
+    setIsModalDismissed(true);
+  }
+
+  useIdleTimer({
+    onIdle: () => {
+      setHasInactiveSessionTimeLimitBeenHit(true);
+      // Have the backend log out in SECONDS_TO_WRAP_UP_AFTER_INACTIVE_SESSION_TIME_LIMIT
+      updateSessionExpiry(
+        new Date().getTime() +
+          SECONDS_TO_WRAP_UP_AFTER_INACTIVE_SESSION_TIME_LIMIT * 1000
+      );
+    },
+    stopOnIdle: true,
+    timeout: inactiveSessionTimeLimitMinutes * 60 * 1000,
+  });
+
+  if (shouldDisplayTimeLimitPrompt(authStatus, now) && !isModalDismissed) {
+    return (
+      <Modal
+        actions={
+          <React.Fragment>
+            <Button variant="primary" onPress={logOut}>
+              Lock Machine Now
+            </Button>
+            <Button onPress={dismissModal}>Dismiss</Button>
+          </React.Fragment>
+        }
+        content={
+          <Prose textCenter>
+            <h1>Session Time Limit</h1>
+            {hasInactiveSessionTimeLimitBeenHit ? (
+              // Inactive session time limit
+              <p>
+                Your session has been inactive for{' '}
+                {pluralize('minutes', inactiveSessionTimeLimitMinutes, true)}.
+                <br />
+                The machine will automatically lock in{' '}
+                <Timer countDownTo={new Date(authStatus.sessionExpiresAt)} />.
+              </p>
+            ) : (
+              // Overall session time limit
+              <p>
+                You are approaching the session time limit of{' '}
+                {pluralize('hours', overallSessionTimeLimitHours, true)}.
+                <br />
+                The machine will automatically lock in{' '}
+                <Timer countDownTo={new Date(authStatus.sessionExpiresAt)} />.
+              </p>
+            )}
+            <p>
+              Lock the machine now and reauthenticate
+              <br />
+              with your smart card to continue working.
```

Comment:
I was on the fence here and thought the text looked a little funny with the natural wrapping as applied by the modal width but also do like the idea of letting things wrap naturally. If it's alright with you, I'm going to leave this as is since I got an LGTM from Roe on the screenshots with this particular wrapping